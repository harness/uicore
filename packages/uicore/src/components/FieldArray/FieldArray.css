/*
 * Copyright 2019 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.container {
  width: 100%;

  .title {
    justify-content: space-between;
    align-items: center;

    .text {
      font-weight: 500;
      padding: 7px 0;
    }
  }

  .noData {
    border: 1px solid var(--grey-300);
    border-radius: 4px;
    text-align: center;
    padding: var(--spacing-large);

    .text {
      line-height: 1.5;
      margin-bottom: 20px;
    }
  }

  table {
    border: 1px solid var(--grey-300);
    border-radius: 4px;
    width: 100%;

    thead {
      th {
        border-bottom: 1px solid var(--grey-300);
        padding: 15px;
        font-weight: normal;
        font-size: smaller;
        text-align: left;
        color: var(--grey-500);
      }
    }

    tbody {
      tr {
        td {
          border-bottom: 1px solid var(--grey-200);
          padding: 10px 15px;
        }

        td.deleteColumn {
          text-align: right;
        }

        input[type='text'] {
          border: 1px solid transparent;
        }

        &:hover {
          background-color: var(--blue-100);

          input[type='text'] {
            border: 1px solid var(--grey-400);
          }
        }
      }
      tr:last-child {
        td {
          border-bottom: none;
        }
      }
    }
  }
}
