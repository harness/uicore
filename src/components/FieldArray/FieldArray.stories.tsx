import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary, Description } from '@storybook/addon-docs/blocks'
import { FieldArray } from '../..'
import FieldArrayExample from './Examples/FieldArrayExample'

export default {
  title: 'Form / FieldArray',

  component: FieldArray,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>FieldArray</Title>
            <Subtitle>
              <>
                <pre>
                  <code>{`import {FieldArray} from '@harness/uicore'`}</code>
                </pre>
                <pre>
                  <code>{`interface Props {
  fields: Field[]
  label: string
  name: string
  placeholder?: string
  isDeleteOfRowAllowed?: (row: React<string, FieldValue>, rowIndex: number) => boolean
  addLabel?: string
  insertRowAtBeginning?: boolean
  onChange?: (params) => void
}

interface Field {
  name: string
  label: string
  defaultValue?: FieldValue
  renderer?: (
    value: FieldValue,
    rowIndex: number,
    onChange: (value: FieldValue>) => void
  ) => React.ReactElement
}

type FieldValue = string | number

type RowData = Record<string, FieldValue>`}</code>
                </pre>

                <h2>Usage</h2>
                <pre>
                  <code>
                    {`const fields: Array<Field> = [
    {
      name: 'col1',
      label: 'Column Header 1',
      defaultValue: 'Item 1'
    },
    {
      name: 'col2',
      label: <div style={{fontWeight: 'bold'}}>Column Header 2</div>,
      defaultValue: 'Item 2',
      renderer: (value, _index, handleChange) => (
        <TextInput
          defaultValue={value}
          placeholder="Column 2 value"
          onChange={e => handleChange(e.target.value)} />
      )
    },
    {
      name: 'col3',
      label: 'Column Header 3',
      defaultValue: 'Item 3',
      renderer: (value, _rowIndex, handleChange) => (
        <MultiSelect
          className="multiSelect"
          items={[
            { label: 'One', value: '1' },
            { label: 'Two', value: '2' }
          ]}
          placeholder="Column 3 value"
          onChange={handleChange}
          value={value || []}
        />
      )
    }
  ]

const data: Array<RowData> = [
  {
      col1: 'row 1 col 1',
      col2: 'row 1 col 2',
      col3: 'row 1 col 3'
  }
]

<Formik initialValues={{ fieldArray: data }}>
  {props => (
    <form onSubmit={props.handleSubmit}>
      <FieldArray name='fieldArray' fields={fields} title="Field List" placeholder="Help Text" />
    </form>
  )}
</Formik>`}
                  </code>
                </pre>
              </>
            </Subtitle>
            <Description></Description>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta
export const Basic: Story<any> = () => {
  return <FieldArrayExample />
}
