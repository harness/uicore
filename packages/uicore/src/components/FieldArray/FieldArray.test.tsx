/*
 * Copyright 2020 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { Formik } from 'formik'

import FieldArray, { Field } from './FieldArray'

const fields: Field[] = [
  {
    name: 'col1',
    label: 'Column Header 1',
    defaultValue: 'Item 1'
  },
  {
    name: 'col2',
    label: 'Column Header 2',
    defaultValue: 'Item 2'
  }
]
const data = [
  {
    col1: 'col 1',
    col2: 'col 2'
  },
  {
    col1: 'col 1',
    col2: 'col 2'
  }
]

const noDataText = 'This is a no data/add data message.'

describe('<FieldArray /> tests', () => {
  test('render placeholder when no initial value given', () => {
    const { container } = render(
      <Formik onSubmit={jest.fn()} initialValues={{}}>
        {props => (
          <form onSubmit={props.handleSubmit}>
            <FieldArray name="fieldArray" fields={fields} label="Field List" placeholder={noDataText} />
          </form>
        )}
      </Formik>
    )
    const noDataTextEl = container.querySelector('.noData .text')

    expect(container).toMatchSnapshot()
    expect(noDataTextEl).not.toBeNull()
    expect(noDataTextEl!.innerHTML).toEqual(noDataText)
  })

  test('render initial value when given', () => {
    const { container } = render(
      <Formik onSubmit={jest.fn()} initialValues={{ fieldArray: data }}>
        {props => (
          <form onSubmit={props.handleSubmit}>
            <FieldArray name="fieldArray" fields={fields} label="Field List" placeholder={noDataText} />
          </form>
        )}
      </Formik>
    )
    expect(container).toMatchSnapshot()
    expect(container.querySelectorAll('tr').length).toEqual(data.length + 1)
  })

  test('should be able to add rows from both buttons', async () => {
    const { container } = render(
      <Formik onSubmit={jest.fn()} initialValues={{}}>
        {props => (
          <form onSubmit={props.handleSubmit}>
            <FieldArray name="fieldArray" fields={fields} label="Field List" placeholder={noDataText} />
          </form>
        )}
      </Formik>
    )
    const btnAddRow = container.querySelector('.noData button[data-id="btn-add-no-data"]')
    fireEvent.click(btnAddRow!)
    await wait()
    expect(container).toMatchSnapshot()
    expect(container.querySelectorAll('tr').length).toEqual(2)
    const btnAddRow2 = container.querySelector('.title button[data-id="btn-add"]')
    fireEvent.click(btnAddRow2!)
    await wait()
    expect(container).toMatchSnapshot()
    expect(container.querySelectorAll('tr').length).toEqual(3)
  })

  test('should be able to delete row', async () => {
    const { container } = render(
      <Formik onSubmit={jest.fn()} initialValues={{ fieldArray: data }}>
        {props => (
          <form onSubmit={props.handleSubmit}>
            <FieldArray name="fieldArray" fields={fields} label="Field List" placeholder={noDataText} />
          </form>
        )}
      </Formik>
    )

    expect(container.querySelectorAll('tr').length).toEqual(3)
    const btnDeleteRow = container.querySelector('tbody tr button[data-id="btn-delete"]')
    fireEvent.click(btnDeleteRow!)
    await wait()
    expect(container).toMatchSnapshot()
    expect(container.querySelectorAll('tr').length).toEqual(2)
  })

  test('should be able to render custom fields', async () => {
    const customField: Field = {
      name: 'col3',
      label: 'Column Header 3',
      defaultValue: 'Item 3',
      // eslint-disable-next-line react/display-name
      renderer: value => <div id="#customField">{value}</div>
    }
    const { container } = render(
      <Formik onSubmit={jest.fn()} initialValues={{ fieldArray: data }}>
        {props => (
          <form onSubmit={props.handleSubmit}>
            <FieldArray
              name="fieldArray"
              fields={fields.concat(customField)}
              label="Field List"
              placeholder={noDataText}
            />
          </form>
        )}
      </Formik>
    )

    expect(container).toMatchSnapshot()
    expect(container.querySelector('#customField')).not.toBeUndefined()
  })
})
