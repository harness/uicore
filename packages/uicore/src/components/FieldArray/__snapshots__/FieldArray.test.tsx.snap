// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldArray /> tests render initial value when given 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color without-shadow withLeftIcon"
          data-id="btn-add"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
      <table
        cellspacing="0"
      >
        <thead>
          <tr>
            <th>
              Column Header 1
            </th>
            <th>
              Column Header 2
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              col 1
            </td>
            <td>
              col 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              col 1
            </td>
            <td>
              col 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </form>
</div>
`;

exports[`<FieldArray /> tests render placeholder when no initial value given 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
      </div>
      <div
        class="noData"
      >
        <div
          class="text"
        >
          This is a no data/add data message.
        </div>
        <button
          class="bp3-button button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color withLeftIcon"
          data-id="btn-add-no-data"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`<FieldArray /> tests should be able to add rows from both buttons 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color without-shadow withLeftIcon"
          data-id="btn-add"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
      <table
        cellspacing="0"
      >
        <thead>
          <tr>
            <th>
              Column Header 1
            </th>
            <th>
              Column Header 2
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Item 1
            </td>
            <td>
              Item 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </form>
</div>
`;

exports[`<FieldArray /> tests should be able to add rows from both buttons 2`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color without-shadow withLeftIcon"
          data-id="btn-add"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
      <table
        cellspacing="0"
      >
        <thead>
          <tr>
            <th>
              Column Header 1
            </th>
            <th>
              Column Header 2
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Item 1
            </td>
            <td>
              Item 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              Item 1
            </td>
            <td>
              Item 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </form>
</div>
`;

exports[`<FieldArray /> tests should be able to delete row 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color without-shadow withLeftIcon"
          data-id="btn-add"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
      <table
        cellspacing="0"
      >
        <thead>
          <tr>
            <th>
              Column Header 1
            </th>
            <th>
              Column Header 2
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              col 1
            </td>
            <td>
              col 2
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </form>
</div>
`;

exports[`<FieldArray /> tests should be able to render custom fields 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <div
        class="horizontal layout-spacing-undefined StyledProps--main title"
      >
        <span
          class="text"
        >
          Field List
        </span>
        <button
          class="bp3-button bp3-minimal button StyledProps--font StyledProps--main StyledProps--intent-primary with-current-color without-shadow withLeftIcon"
          data-id="btn-add"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
            icon="plus"
          >
            <svg
              data-icon="plus"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                plus
              </desc>
              <path
                d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Add
          </span>
        </button>
      </div>
      <table
        cellspacing="0"
      >
        <thead>
          <tr>
            <th>
              Column Header 1
            </th>
            <th>
              Column Header 2
            </th>
            <th>
              Column Header 3
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              col 1
            </td>
            <td>
              col 2
            </td>
            <td>
              <div
                id="#customField"
              />
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              col 1
            </td>
            <td>
              col 2
            </td>
            <td>
              <div
                id="#customField"
              />
            </td>
            <td
              class="deleteColumn"
            >
              <button
                class="bp3-button bp3-minimal button StyledProps--font StyledProps--main with-current-color iconOnly without-shadow withLeftIcon"
                data-id="btn-delete"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="main-trash"
                >
                  <svg
                    height="16"
                    viewBox="0 0 16 20"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M1.143 8.48a1.2 1.2 0 011.189-1.36h10.649a1.2 1.2 0 011.194 1.321l-1.039 10.208a1.2 1.2 0 01-1.193 1.079H3.716a1.2 1.2 0 01-1.189-1.04L1.143 8.48zm11.837-.16H2.332l1.385 10.207h8.226L12.98 8.32zM5.916 1.594a.6.6 0 01.719-.452l4.224.965a.6.6 0 01.452.719L11 4.179l3.86.882a.6.6 0 01-.267 1.17L1.535 3.248a.6.6 0 11.267-1.17l3.805.87.309-1.354zm.86 1.62l3.056.698.175-.768-3.055-.698-.175.768z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M8 10.808a.5.5 0 01.5.5v5a.5.5 0 01-1 0v-5a.5.5 0 01.5-.5zm-2.761.012a.5.5 0 01.541.454l.436 4.981a.5.5 0 11-.996.087l-.436-4.98a.5.5 0 01.455-.542zM10.762 10.82a.5.5 0 01.455.541l-.436 4.981a.5.5 0 11-.996-.087l.435-4.98a.5.5 0 01.542-.455z"
                      fill="currentColor"
                      fill-rule="evenodd"
                      opacity="0.3"
                    />
                  </svg>
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </form>
</div>
`;
